types/trigger.d.ts

```diff
@@ -1,10 +1,16 @@
 import { Lang } from './global';
 
-export type Matches = {
-  target?: string;
-  [s: string]: string;
+export interface BaseRegExp<T> extends RegExp {
+  groups?: {
+    [s in T]?: string;
+  };
 }
 
+export type Matches<T> =
+  T extends BaseRegExp ? T['groups'] :
+  T extends RegExp ? { [s: string]: string } :
+  unknown;
```

Comment:
When is T not a BaseRegExp or a RegExp?

Reply:
Yep, when T is not either a BaseRegExp nor RegExp, then return an `unknown`.
Or we can return a `never` so that `tsc` would report an error that you have written a wrong generic type.